python/ql/src/semmle/python/ApiGraphs.qll

```diff
@@ -0,0 +1,417 @@
+/**
+ * Provides an implementation of  _API graphs_, which are an abstract representation of the API
+ * surface used and/or defined by a code base.
+ *
+ * The nodes of the API graph represent definitions and uses of API components. The edges are
+ * directed and labeled; they specify how the components represented by nodes relate to each other.
+ */
+
+import python
+import semmle.python.dataflow.new.DataFlow
+
+/**
+ * Provides classes and predicates for working with APIs used in a database.
+ */
+module API {
+  /**
+   * An abstract representation of a definition or use of an API component such as a function
+   * exported by a Python package, or its result.
+   */
+  class Node extends Impl::TApiNode {
+    /**
+     * Gets a data-flow node corresponding to a use of the API component represented by this node.
+     *
+     * For example, `import re; re.escape` is a use of the `escape` function from the
+     * `re` module, and `import re; re.escape("hello")` is a use of the return of that function.
+     *
+     * This includes indirect uses found via data flow, meaning that in
+     * ```python
+     * def f(x):
+     *     pass
+     *
+     * f(obj.foo)
+     * ```
+     * both `obj.foo` and `x` are uses of the `foo` member from `obj`.
+     */
+    DataFlow::Node getAUse() {
+      exists(DataFlow::LocalSourceNode src | Impl::use(this, src) |
+        Impl::trackUseNode(src).flowsTo(result)
+      )
+    }
+
+    /**
+     * Gets an immediate use of the API component represented by this node.
+     *
+     * For example, `import re; re.escape` is a an immediate use of the `escape` member
+     * from the `re` module.
+     *
+     * Unlike `getAUse()`, this predicate only gets the immediate references, not the indirect uses
+     * found via data flow. This means that in `x = re.escape` only `re.escape` is a reference
+     * to the `escape` member of `re`, neither `x` nor any node that `x` flows to is a reference to
+     * this API component.
+     */
+    DataFlow::LocalSourceNode getAnImmediateUse() { Impl::use(this, result) }
+
+    /**
+     * Gets a call to the function represented by this API component.
+     */
+    DataFlow::Node getACall() { result = getReturn().getAnImmediateUse() }
```

Comment:
I don't understand the second part of your comment, but to answer the first I'll note that there _are_ external uses of `getAnImmediateUse` in the JS libs, so I imagine it will come in handy for us as well.

Reply:
Ok, for the first part, then :-) For the second, it just felt a bit brittle to assume that any immediate use of a return would be a call. It looks to be the case by the definition of returns further down, but if that definition gets expanded, does this logic follow along? That to explain my thoughts; it is probably fine to follow javascript here for now.

Reply:
This is a fair point, but I don't have an answer other than "let's be really careful when we add `return`-labelled edges". :slightly_smiling_face: